doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,189 @@
+basic need, explicitly fetched signature algorithms.  The following method
+type is affected by this document:
+
+- `EVP_SIGNATURE`
+
+Public API - Add variants of `EVP_PKEY_CTX` functionality
+---------------------------------------------------------
+
+Through OTC discussions, it's been determined that the most suitable APIs to
+touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
+`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
+associated functions.  They can be extended to accept an explicitly fetched
+algorithm of the right type, and to be able to process infinite amounts of
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
+
+It must be made clear that the added functionality can not be used to
+compose an algorithm from different parts.  For example, it's not possible
+to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
+specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
+`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
+other input that's limited to the modulus size of the RSA pkey.
+
+### Making things less confusing with distinct function names
+
+So far, `EVP_PKEY_sign()` and friends are only expected to act as
+"primitive" functions.  This design essentially proposes an extension to
+also allow streaming functionality through an *update* and a *final*
+function.  Discussions have revealed that it is potentially confusing to
+confound pure "primitive" functionality with functionality that expects a
+digest as input, as well as with "streaming" functionality into the same
+name, so this design also includes a naming proposal that distinguishes
+the two uses.
+
+### Making it possible to verify with an early signature
+
+There are newer verifying algorithms that need to receive the signature
+before processing the data.  This is particularly important for streaming
+functionality.  This design proposes a mechanism to accomodate this, which
+is different than the mechanism you get with `EVP_PKEY_verify()`.
+
+Public API - API Reference
+--------------------------
+
+### For signing with `EVP_SIGNATURE`
+
+#### Signing a digest
+
+``` C
+int EVP_PKEY_sign_init_for_digest(EVP_PKEY_CTX *pctx,
+                                  EVP_SIGNATURE *algo,
+                                  const OSSL_PARAM params[]);
+```
+
+(notably, a fetched "RSA" should probably not accept `RSA_NO_PADDING` when
+initialized with this call)
+
+#### Signing a stream
+
+``` C
+int EVP_PKEY_sign_init_for_message(EVP_PKEY_CTX *pctx,
+                                   EVP_SIGNATURE *algo,
+                                   const OSSL_PARAM params[]);
+int EVP_PKEY_sign_update(EVP_PKEY_CTX *ctx,
+                         const unsigned char *in, size_t inlen);
+int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,
+                        unsigned char *sig, size_t *siglen);
+```
```

Comment:
We will also need a oneshot. I.e., EVP_PKEY_sign_message(). And I would change those two above to EVP_PKEY_sign_message_update() and EVP_PKEY_sign_message_final().